pkg/controller/machinedeployment/controller.go

```diff
@@ -15,41 +14,328 @@ See the License for the specific language governing permissions and
+	msList, err := c.msLister.MachineSets(d.Namespace).List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: flush out machine set adoption.
+
+	var filteredMS []*v1alpha1.MachineSet
+	for _, ms := range msList {
+		if metav1.GetControllerOf(ms) != nil && !metav1.IsControlledBy(ms, d) {
+			glog.V(4).Infof("%s not controlled by %v", ms.Name, d.Name)
+			continue
+		}
+		selector, err := metav1.LabelSelectorAsSelector(&d.Spec.Selector)
+		if err != nil {
+			continue
+		}
+		// If a deployment with a nil or empty selector creeps in, it should match nothing, not everything.
+		if selector.Empty() || !selector.Matches(labels.Set(ms.Labels)) {
+			continue
+		}
+		filteredMS = append(filteredMS, ms)
+	}
+	return filteredMS, nil
+}
+
+// Reconcile handles reconciling of machine deployment
 func (c *MachineDeploymentControllerImpl) Reconcile(u *v1alpha1.MachineDeployment) error {
-	// Implement controller logic here
-	log.Printf("Running reconcile MachineDeployment for %s\n", u.Name)
-	return nil
+	// Deep-copy otherwise we are mutating our cache.
+	d := u.DeepCopy()
+
+	everything := metav1.LabelSelector{}
+	if reflect.DeepEqual(d.Spec.Selector, &everything) {
+		if d.Status.ObservedGeneration < d.Generation {
+			d.Status.ObservedGeneration = d.Generation
+			c.machineClient.ClusterV1alpha1().MachineDeployments(d.Namespace).UpdateStatus(d)
+		}
+		return nil
+	}
+
+	msList, err := c.getMachineSetsForDeployment(d)
+	if err != nil {
+		return err
+	}
+
+	machineMap, err := c.getMachineMapForDeployment(d, msList)
+	if err != nil {
+		return err
+	}
+
+	if d.DeletionTimestamp != nil {
+		return c.sync(d, msList, machineMap)
+	}
+
+	if d.Spec.Paused {
+		return c.sync(d, msList, machineMap)
+	}
+
+	switch d.Spec.Strategy.Type {
+	case common.RollingUpdateMachineDeploymentStrategyType:
+		return c.rolloutRolling(d, msList, machineMap)
+	}
+	return fmt.Errorf("unexpected deployment strategy type: %s", d.Spec.Strategy.Type)
 }
 
 func (c *MachineDeploymentControllerImpl) Get(namespace, name string) (*v1alpha1.MachineDeployment, error) {
-	return c.lister.MachineDeployments(namespace).Get(name)
+	return c.mdLister.MachineDeployments(namespace).Get(name)
+}
+
+// addMachineSet enqueues the deployment that manages a MachineSet when the MachineSet is created.
+func (c *MachineDeploymentControllerImpl) addMachineSet(obj interface{}) {
+	ms := obj.(*v1alpha1.MachineSet)
+
+	if ms.DeletionTimestamp != nil {
+		// On a restart of the controller manager, it's possible for an object to
+		// show up in a state that is already pending deletion.
+		c.deleteMachineSet(ms)
+		return
+	}
+
+	// If it has a ControllerRef, that's all that matters.
+	if controllerRef := metav1.GetControllerOf(ms); controllerRef != nil {
+		d := c.resolveControllerRef(ms.Namespace, controllerRef)
+		if d == nil {
+			return
+		}
+		glog.V(4).Infof("MachineSet %s added for deployment %v.", ms.Name, d.Name)
+		c.Reconcile(d)
```

Comment:
Instead of calling Reconcile directly, you should queue the MachineDeployment. Otherwise, you could have Reconcile called simultaneously on the same MachineDeployment. (Same applies to all calls to Reconcile from event handlers)

Reply:
You can get access to the queue in your `Init` function with:
```go
queue := arguments.GetSharedInformers().WorkerQueues["MachineDeployment"].Queue
```